text/0000-unsafe-attributes.md

```diff
@@ -0,0 +1,175 @@
+  to opt-in to a soundness-critical diagnostic -- that's totally against the
+  "safety by default" goal of Rust). This RFC says that yes, `deny(unsafe_code)`
+  should deny those attributes, but we should go further and require an explicit
+  `unsafe(...)` attribute block for them to be used at all.
+- **Item-level unsafe blocks.** We could find some way to have 'unsafe blocks'
+  around entire functions or modules. However, those would go against the usual
+  goal of keeping `unsafe` blocks small. Big `unsafe` blocks risk accidentally
+  calling an unsafe operation in there without even realizing it.
+- **Other syntax.** Obviously we could pick a different syntax for the same
+  concept, but this seems like the most natural marriage of the idea of unsafe
+  blocks from regular code, and the existing attributes syntax.
+
+# Prior art
+[prior-art]: #prior-art
+
+We have `unsafe` blocks; this is basically the same thing for the "attributes
+DSL".
+
+In the attribute DSL, we already have a "nesting" construct: `cfg_attr`. That
+allows terms like
+`#[cfg_attr(debug_assertions, deny(unsafe_code), allow(unused))]`, so there is
+precedent for having a list of attributes inside a single attribute.
+
+I don't know of other languages that would distinguish safe and unsafe
+attributes.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- **Different lint staging.** The lint on using existing unsafe attributes like
+  `no_mangle` outside `unsafe(...)` could be staged in various ways: it could be
+  warn-by-default to start or we wait a while before to do that, it could be
+  edition-dependent, it might eventually be deny-by-default or even a hard error
+  on some editions -- there are lots of details here, which can be determined
+  later during the process.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- **Unsafe attribute proc macros.** We could imagine something like
+    ```
+    #[proc_macro_attribute(require_unsafe)]
+    fn spoopy(args: TokenStream, input: TokenStream) -> TokenStream {…}
+    ```
+  to declare that an attribute proc macro is unsafe to use, and must only
+  occur as an unsafe macro. Such an unsafe-to-use attribute proc macro must
+  declare in a comment what its safety requirements are. (This is the `unsafe`
+  from `unsafe fn`, whereas the rest of the RFC is using the `unsafe` from
+  `unsafe { ... }`.)
+- **Unsafe derive.** We could use `#[unsafe(derive(Trait))]` to derive an
+  `unsafe impl` where the deriving macro itself cannot check all required safety
+  conditions.
+- **Unsafe tool attributes.** Same as above, but for tool attributes.
+- **Unsafe attributes on statements.** For now, the only unsafe attributes we
+  have don't make sense on the statement level. Once we do have unsafe statement
+  attributes, we need to figure out whether inside `unsafe {}` blocks one still
+  needs to also write `unsafe(...)`.
```

Comment:
We do have to settle this. You can put declarations inside a block within a function, including AFAIK an unsafe block.

Reply:
A `fn` inside of an `unsafe` block inside of another `fn` doesn't "inherit" the unsafe block over its body, so it would be most logical for it to also not inherit the unsafe block over its attributes.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59b2b073cba046f252f19218933b0d21

Reply:
Good point, will add a note on that.